river/cluster/clustream.py

```diff
@@ -0,0 +1,363 @@
+    ...     clustream_test = clustream_test.learn_one(x)
+
+    >>> clustream_test.predict_one({0: 1, 1: 1})
+    1
+
+    >>> clustream_test.predict_one({0: 4, 1: 3})
+    0
+
+    References
+    ----------
+    .. [1] A. Kumar, A. Singh, and R. Singh, 2017. An efficient hybrid-clustream algorithm
+       for stream mining. In Proceedings of the 13th International Conference on Signal-Image Technology & Internet-Based System (SITIS).
+       DOI: 10.1109/SITIS.2017.77
+
+    """
+
+    def __init__(
+        self,
+        seed: int = None,
+        time_window: int = 1000,
+        max_kernels: int = 100,
+        kernel_radius_factor: int = 2,
+        number_of_clusters: int = 5,
+        **kwargs
+    ):
+        super().__init__()
+        self.time_window = time_window
+        self.time_stamp = -1
+        self.kernels = {n: None for n in range(max_kernels)}
+        self.initialized = False
+        self.buffer = {}
+        self.buffer_size = max_kernels
+        self.centers = (
+            {}
+        )  # add this to retrieve centers later for the evaluation of models through
+        self.kernel_radius_factor = kernel_radius_factor
+        self.max_kernels = max_kernels
+        self.number_of_clusters = number_of_clusters
+        self._train_weight_seen_by_model = 0.0
+        self.seed = seed
+        self.kwargs = kwargs
+
+    def learn_one(self, x, sample_weight=None):
+        """Incrementally trains the model.
+
+        Tasks performed before training:
+        * Verify instance weight. if not provided, uniform weights (1.0) are assumed.
+        * Update weight seen by model.
+
+        Training tasks:
+        * determinate closest kernel
+        * Check whether instance fits into closest Kernel:
+            1- if data fits, put into kernel
+            2- if data does not fit, we need to free some space to insert a new kernel
+            and this can be done in two ways, delete an old kernel or merge two kernels
+            which are close to each other
+
+        Parameters
+        ----------
+        x
+            A dictionary of features
+
+        sample_weight
+            Instance weights. If not provided, uniform weights are assumed
+
+        Returns
+        ----------
+        self
+
+        """
+        if sample_weight == 0:
+            return
+        elif sample_weight is None:
+            sample_weight = 1.0
+
+        self._train_weight_seen_by_model += sample_weight
+
+        # merge _learn_one into learn_one
+        self.time_stamp += 1
+
+        if not self.initialized:
+            if len(self.buffer) < self.buffer_size:
+                self.buffer[len(self.buffer)] = ClustreamKernel(
+                    x=x,
+                    sample_weight=sample_weight,
+                    timestamp=self.time_stamp,
+                    T=self.kernel_radius_factor,
+                    M=self.max_kernels,
+                )
+                return self
+            else:
```

Comment:
No need for the `else` here because there's a `return` in the `if`. In general I think it's a good idea to try to un-nest logic as much as possible.

Reply:
Resolved. Also separated it to a new function called `initialise_learn_one`.